### PR TITLE
always realize output regardless of the number of reads

### PR DESCRIPTION
**Pull request by ngimel**
This improves hf_Bert 1.139x->1.21x, currently lowmem dropout doesn't work for nn.Dropout module, and before this change we were recomputing all the dropout masks in a very inefficient kernel. This change pushes dropout masks to be saved in the dropout kernels where they are first computed. 

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

### CODE REVIEW

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -3505,11 +3505,13 @@ def realize(self):
         self.data.name = V.graph.register_buffer(self.data)
         return self.data.name
 
-    def realize_hint(self):
+    def realize_hint(self, is_output=False):
         """
         Called on buffers we expect to be forced to realize later.
         """
-        if isinstance(self.data, (Pointwise, Reduction)) and self.num_reads() > 1:
```
Comment (Chillee) [inline torch/_inductor/ir.py:3512]: Wait.... why is this `num_reads`? Why would we want to avoid realizing a *buffer* that has 1 read? Shouldn't we want to avoid realizing a buffer that is read *from* only once?

File: torch/_inductor/graph.py [hunk 2]
```diff
@@ -304,7 +304,8 @@ def run_node(self, n: torch.fx.Node):
             if num_users > 1 and isinstance(result, TensorBox):
                 for user in n.users:
                     if user.target in needs_realized_inputs or user.op == "output":
```
Comment (Chillee) [inline torch/_inductor/graph.py:306]: I think it'd probably be cleaner if we just added a separate case here.

So something like
```
if user.target in needs_realized_inputs:
    result.realize_hint()
elif user.op == 'output':
    result.realize()
```

Comment (ngimel) [reply torch/_inductor/graph.py:306]: We'd still need a check of `isinstance(self.data, (Pointwise, Reduction))` here, but sure, I can move it here.

File: torch/_inductor/ir.py (hunk 1 above)
Comment (ngimel) [reply torch/_inductor/ir.py]: I guess we want to avoid realizing it because if it's just 1 read, recomputing or realizing is the same in terms of global memory reads (not quite, due to potentially different dtype, but close enough). It's all part of our recompute/realize heuristics that's pretty arbitrary, but some things are unambiguous, like once you computed output, write it out, you have to anyway.

**Review by Chillee — APPROVED** [review]

**Comment by Chillee** [comment]
Would really be nice to have one of those deterministic memory-bandwidth tests :)

File: torch/_inductor/graph.py (hunk 2 above)
Comment (jgong5) [reply torch/_inductor/graph.py]: Is there a case where we don't want to realize a buffer having the 'output' as the user? I saw the `num_users > 1` check here, not sure if it is needed for the output..

File: torch/_inductor/scheduler.py [hunk 3]
```diff
@@ -922,9 +922,22 @@ def can_fuse_vertical(self, node1, node2):
         be scheduled before the fusion of node1 and node2.
         """
         node1_names = node1.get_names()
-        remaining_deps = {
-            dep.name for dep in node2.unmet_dependencies - node1.read_writes.writes
-        }
+        computed_deps = set()
+        for rd in node2.unmet_dependencies:
```
Comment (Chillee) [inline torch/_inductor/scheduler.py:926]: Could we normalize this some other way? 

Like, the logic here seems like something that's generically true about `MemoryDep`?

Comment (ngimel) [reply torch/_inductor/scheduler.py:926]: I'll leave that for the future, we could create a comparison op for MemoryDeps that would return True for this.
